### PR TITLE
feat!: Deprecate pipeline code and allow to select the target pipeline when pushing(HEXA-1170)

### PR DESCRIPTION
**Pull request by YolanFery**
We want to decouple the pipeline code from the python code

## Changes

- `select_pipeline` routine to ask for which pipeline to update
- Deprecate `code` in pipeline definition and show a warning message
- Extend the push test and add 3 unit test on the sub method to cover edge cases

## How/what to test

Pushing pipelines with all the options

## Screenshots / screencast

Screenshot 2025-02-19 at 15 57 28

### CODE REVIEW

File: tests/test_cli.py [hunk 1]
```diff
@@ -148,13 +158,17 @@ def test_push_pipeline(self, mock_create_template, mock_upload_pipeline, mock_ge
             mock_create_template.return_value = template
 
             result = self.runner.invoke(
-                pipelines_push, [tmp, "--name", version], input="\n".join(["Y", "Y", changelog]) + "\n"
+                pipelines_push,
+                [tmp, "--name", version],
+                input="\n".join(["4", "pipeline_code", "Y", "Y", changelog]) + "\n",
             )
             self.assertEqual(result.exit_code, 0)
+            self.assertIn("Which pipeline do you want to update?", result.output)
```
Comment (YolanFery) [inline tests/test_cli.py:166]: Not totally satisfied with the resulting behaviour, I feel like this brings a lot of friction, WDYT @qgerome @nazarfil

File: openhexa/sdk/pipelines/pipeline.py [hunk 2]
```diff
@@ -275,6 +276,13 @@ def pipeline(
     ... def a_task() -> int:
     ...     return 42
     """
+    if code:
+        warnings.warn(
+            f"The pipeline 'code' parameter (with value {code}) is deprecated and ignored and will be removed in future versions. The pipeline code is now auto-generated.",
+            DeprecationWarning,
+        )
```
Comment (YolanFery) [inline openhexa/sdk/pipelines/pipeline.py:283]: Where is this actually printed ?

File: tests/test_cli.py (hunk 1 above)
Comment (YolanFery) [reply tests/test_cli.py:169]: Here is the friction I am thinking of : 
In the current state in production, pushing a pipeline only requires confirmation.
But now we propose multiples choices, so it takes virtually longer to make a choice and push pipeline

File: openhexa/cli/cli.py [hunk 3]
```diff
@@ -260,6 +261,53 @@ def propose_to_create_template_version(workspace, pipeline_version, yes):
...
+                    click.echo(
+                        f"Pipeline with code {click.style(pipeline_code, italic=True)} not found. Please try again."
+                    )
+            else:
+                return workspace_pipelines[choice_idx]
+        elif workspace_pipelines:
+            return workspace_pipelines[0]
+        else:
+            return None
+
```
[the remark this answers is not in the document]
Comment (YolanFery) [reply openhexa/cli/cli.py:313]: Thanks, I agree. 

I now have reworked this part of the code to split it and make it more readable

Not sure about the naming at this point but yes we can revisit

File: openhexa/cli/api.py [hunk 4]
```diff
@@ -169,14 +169,15 @@ def get_workspace(slug: str, token: str):
     )["workspace"]
 
 
-def list_pipelines():
-    """List all pipelines in the workspace."""
+def list_pipelines_pages(name=None):
```
Comment (qgerome) [inline openhexa/cli/api.py:172]:
```suggestion
def list_pipelines_pages(name:str|None = None):
```

Comment (qgerome) [reply openhexa/cli/api.py:172]: Personally I'd change the method name to imply a lower level call like `fetch_pipelines` or `get_pipelines`. To protect also from a change on the backend, I'd specifically limit the pageSize to a set amount.

Comment (qgerome) [reply openhexa/cli/api.py:172]: Now that I see the code, I'm wondering if not passing page or perPage to the query do not just return all the items 🤔

File: openhexa/cli/cli.py [hunk 5]
```diff
@@ -260,6 +261,57 @@ def propose_to_create_template_version(workspace, pipeline_version, yes):
             )
 
 
+def select_pipeline(workspace_pipelines, number_of_pages: int, pipeline):
```
Comment (qgerome) [inline openhexa/cli/cli.py:264]: Nice structure, easy to read :)

File: openhexa/sdk/pipelines/pipeline.py (hunk 2 above)
Comment (qgerome) [reply openhexa/sdk/pipelines/pipeline.py:283]: It's printed when the pipeline is ran locally or in the cloud. So it's not really interesting to put the check here IMO

Comment (qgerome) [reply openhexa/sdk/pipelines/pipeline.py:283]: ```return Pipeline(code, name, fun, parameters, timeout)```

Should be updated as well to remove the code from the `Pipeline` object no ?

**Review by qgerome — CHANGES_REQUESTED** [review]
I think we could adapt the `Pipeline` object to remove the notion of code (this object is not exposed to the end-user)

File: openhexa/cli/api.py (hunk 4 above)
Comment (YolanFery) [reply openhexa/cli/api.py]: Thanks, I have renamed the methods, added comments and provided default for the page sizes. I think that it makes sense to keep a paged number of results to avoid unintentional misuse of the api resulting in big queries.

File: openhexa/cli/cli.py (hunk 5 above)
Comment (YolanFery) [reply openhexa/cli/cli.py:264]: Thanks, the second iteration is definitely better thanks to the suggestions from @nazarfil

**Comment by YolanFery** [comment]
Thanks for the review. I have removed the warnings and refactored `Pipeline` to not accept and rely on code anymore

**Review by qgerome — APPROVED** [review]
